File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    stack = []
    
    for i in range(len(s)):
        
        if(not stack):
            stack.append(s[i])
        elif(stack and stack[-1] == '('): # 얘를 if로 분기하면 안됨...실수주의
            if(s[i] == '('):
                stack.append(s[i])
            else: # s[i] == ')'
                stack.pop()
            
    if(not stack):
        return True
    else:
        return False
    
    

        

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
// non-stack 버전
class Solution {
    boolean solution(String s) {

        int cnt = 0;
        for(Character c : s.toCharArray()) {
            if(c == '(') {
                cnt++;
            } else {
                cnt--;
            }
            if(cnt < 0) {
                return false;
            }
        }

        return cnt == 0;
    }
}

/*
    가장 최신의 상태만 추적하면 되므로
    스택 없이 상수값으로도 확인할 수 있다.

    만약 open이 들어오면 cnt + 1
    close가 들어오면 cnt - 1
    
    이렇게 해서 최종 순간에 cnt가 0이 되면 올바른 쌍만 들어왔다는 뜻

    근데 이 경우
    )( 같은 케이스는 cnt가 0이되는데
    )가 먼저 들어오는 순간에는 cnt<0 이 될 것이므로
    그냥 바로 false쳐버림
*/